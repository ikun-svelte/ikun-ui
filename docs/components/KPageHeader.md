---
title: KPageHeader
lang: en-US
---

# KPageHeader

If path of the page is simple, it is recommended to use PageHeader instead of the Breadcrumb.

## Install

::: code-group

```bash [pnpm]
pnpm add @ikun-ui/page-header
```

```bash [yarn]
yarn add @ikun-ui/page-header
```

```bash [npm]
npm install @ikun-ui/page-header
```

:::

## Basic usage

<demo src="page-header/basic.svelte"  github='Pageheader'></demo>

## Additional operation section

The header can be as complicated as needed,
you may add additional sections to the header, to allow rich interactions.

<demo src="page-header/extra.svelte"  github='Pageheader'></demo>

## Main content

Sometimes we want the head to show with some co-responding content,
we can utilize the `default` slot for doing so.

<demo src="page-header/main.svelte"  github='Pageheader'></demo>

## Breadcrumbs

Page header allows you to add breadcrumbs for giving route information to the users by `breadcrumb` slot.

<demo src="page-header/breadcrumbs.svelte"  github='Pageheader'></demo>

## Custom icon

You can customize the icon through the slot or icon attribute

<demo src="page-header/icon.svelte"  github='Pageheader'></demo>

## No icon

PageHeader without icon is also possible

<demo src="page-header/no-icon.svelte"  github='Pageheader'></demo>

## Complete example

<demo src="page-header/complete.svelte"  github='Pageheader'></demo>

## PageHeader Props

| Name    | Type     | Default               | Description                                                 |
| ------- | -------- | --------------------- | ----------------------------------------------------------- |
| icon    | `string` | `i-carbon-arrow-left` | Icon component of page header.following the unocss standard |
| title   | `string` | `''`                  | Title of page header.                                       |
| content | `string` | `''`                  | Content of page header.                                     |
| cls     | `string` | `''`                  | Additional class for component                              |
| attrs   | `any`    | `{}`                  | Additional attributes                                       |

## PageHeader Events

| Name | Description                          | Type        |
| ---- | ------------------------------------ | ----------- |
| back | Triggers when right side is clicked. | `()=> void` |

## PageHeader Slots

| Name       | Description            |
| ---------- | ---------------------- |
| default    | Main content.          |
| title      | Content of the title.  |
| icon       | Content of the icon.   |
| content    | Content.               |
| extra      | Extra.                 |
| breadcrumb | Content as breadcrumb. |
