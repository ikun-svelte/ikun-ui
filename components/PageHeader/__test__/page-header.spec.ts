import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';
import KPageHeader from '../src';
import KPageHeaderIcon from './fixture/slot.icon.svelte';
import KPageHeaderBreadcrumb from './fixture/slot.breadcrumb.svelte';
import KPageHeaderExtra from './fixture/slot.extra.svelte';
import KPageHeaderMain from './fixture/slot.default.svelte';
import KPageHeaderContent from './fixture/slot.content.svelte';
import KPageHeaderTitle from './fixture/slot.title.svelte';
import { tick } from 'svelte';
let host;

const initHost = () => {
	host = globalThis.document.createElement('div');
	host.setAttribute('id', 'host');
	globalThis.document.body.appendChild(host);
};
beforeEach(() => {
	initHost();
	vi.useFakeTimers();
});
afterEach(() => {
	host.remove();
	vi.useRealTimers();
});

const CONTENT = 'Iâ€˜m Ikun';
describe('Test: KPageHeader', () => {
	test('props: cls', async () => {
		// @ts-ignore
		const instance = new KPageHeader({
			target: host,
			props: {
				cls: 'k-page-header--test'
			}
		});
		expect(instance).toBeTruthy();
		expect(host!.innerHTML.includes('k-page-header--test')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('basic render', async () => {
		// @ts-ignore
		const instance = new KPageHeader({
			target: host,
			props: {
				content: CONTENT
			}
		});
		expect(instance).toBeTruthy();
		const contentElm = host.querySelector('.k-page-header__content');
		expect(contentElm.innerHTML).toBe(CONTENT);
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: icon', async () => {
		// @ts-ignore
		const instance = new KPageHeader({
			target: host,
			props: {
				icon: 'i-carbon-logo-vue'
			}
		});
		expect(instance).toBeTruthy();
		expect(
			host.innerHTML.includes(
				'class="k-icon--base k-icon--base__dark k-icon-transition i-carbon-logo-vue"'
			)
		).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: title', async () => {
		// @ts-ignore
		const instance = new KPageHeader({
			target: host,
			props: {
				title: CONTENT
			}
		});
		expect(instance).toBeTruthy();
		const titleElm = host.querySelector('.k-page-header__title');
		expect(titleElm.innerHTML).toBe(CONTENT);
		expect(host.innerHTML).matchSnapshot();
	});

	test('slots: icon', async () => {
		// @ts-ignore
		const instance = new KPageHeaderIcon({
			target: host
		});
		expect(instance).toBeTruthy();
		expect(host.innerHTML.includes('<span slot="icon">custom icon</span>')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('slots: title', async () => {
		// @ts-ignore
		const instance = new KPageHeaderTitle({
			target: host
		});
		expect(instance).toBeTruthy();
		expect(host.innerHTML.includes('<span slot="title">custom title</span>')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('slots: extra', async () => {
		// @ts-ignore
		const instance = new KPageHeaderExtra({
			target: host
		});
		expect(instance).toBeTruthy();
		expect(host.innerHTML.includes('<span slot="extra">custom extra</span>')).toBeTruthy();
		expect(host.innerHTML.includes('k-page-header--has-extra')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('slots: default', async () => {
		// @ts-ignore
		const instance = new KPageHeaderMain({
			target: host
		});
		expect(instance).toBeTruthy();
		expect(host.innerHTML.includes('<span>custom default</span>')).toBeTruthy();
		expect(host.innerHTML.includes('k-page-header--contentful')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('slots: content', async () => {
		// @ts-ignore
		const instance = new KPageHeaderContent({
			target: host
		});
		expect(instance).toBeTruthy();
		expect(
			host.innerHTML.includes(
				'<div class="k-page-header__content"><span slot="content">custom content</span></div>'
			)
		).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('slots: breadcrumb', async () => {
		// @ts-ignore
		const instance = new KPageHeaderBreadcrumb({
			target: host
		});
		expect(instance).toBeTruthy();
		expect(
			host.innerHTML.includes(
				'<div class="k-page-header__breadcrumb"><span slot="breadcrumb">custom breadcrumb</span></div>'
			)
		).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('event: back', async () => {
		const mockFn = vi.fn();
		// @ts-ignore
		const instance = new KPageHeader({
			target: host,
			props: {
				content: CONTENT
			}
		});
		expect(instance).toBeTruthy();
		// @ts-ignore
		instance.$on('back', () => {
			mockFn();
		});
		const backElm = host.querySelector('.k-page-header__back');
		backElm.click();
		await tick();
		expect(mockFn).toBeCalled();
	});
});
