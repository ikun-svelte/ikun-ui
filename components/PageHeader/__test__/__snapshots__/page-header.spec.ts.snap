// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test: KPageHeader > basic render 1`] = `"<div class=\\"k-page-header\\"> <div class=\\"k-page-header__head\\"><div class=\\"k-page-header__left\\"><div class=\\"k-page-header__back\\" role=\\"button\\" tabindex=\\"0\\" aria-hidden=\\"true\\"><div aria-label=\\"Back\\" class=\\"k-page-header__icon\\"><span class=\\"k-icon--base k-icon--base__dark k-icon-transition i-carbon-arrow-left\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 16px; height: 16px;\\"></span></div> <div class=\\"k-page-header__title\\">Back</div></div> <div class=\\"k-divider k-divider--vertical k-divider--vertical__solid k-divider--content-center b-ikun-bd-base\\" data-divider=\\"k-divider\\"></div> <div class=\\"k-page-header__content\\">I‘m Ikun</div></div> </div> </div>"`;

exports[`Test: KPageHeader > props: cls 1`] = `"<div class=\\"k-page-header k-page-header--test\\"> <div class=\\"k-page-header__head\\"><div class=\\"k-page-header__left\\"><div class=\\"k-page-header__back\\" role=\\"button\\" tabindex=\\"0\\" aria-hidden=\\"true\\"><div aria-label=\\"Back\\" class=\\"k-page-header__icon\\"><span class=\\"k-icon--base k-icon--base__dark k-icon-transition i-carbon-arrow-left\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 16px; height: 16px;\\"></span></div> <div class=\\"k-page-header__title\\">Back</div></div> <div class=\\"k-divider k-divider--vertical k-divider--vertical__solid k-divider--content-center b-ikun-bd-base\\" data-divider=\\"k-divider\\"></div> <div class=\\"k-page-header__content\\"></div></div> </div> </div>"`;

exports[`Test: KPageHeader > props: icon 1`] = `"<div class=\\"k-page-header\\"> <div class=\\"k-page-header__head\\"><div class=\\"k-page-header__left\\"><div class=\\"k-page-header__back\\" role=\\"button\\" tabindex=\\"0\\" aria-hidden=\\"true\\"><div aria-label=\\"Back\\" class=\\"k-page-header__icon\\"><span class=\\"k-icon--base k-icon--base__dark k-icon-transition i-carbon-logo-vue\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 16px; height: 16px;\\"></span></div> <div class=\\"k-page-header__title\\">Back</div></div> <div class=\\"k-divider k-divider--vertical k-divider--vertical__solid k-divider--content-center b-ikun-bd-base\\" data-divider=\\"k-divider\\"></div> <div class=\\"k-page-header__content\\"></div></div> </div> </div>"`;

exports[`Test: KPageHeader > props: title 1`] = `"<div class=\\"k-page-header\\"> <div class=\\"k-page-header__head\\"><div class=\\"k-page-header__left\\"><div class=\\"k-page-header__back\\" role=\\"button\\" tabindex=\\"0\\" aria-hidden=\\"true\\"><div aria-label=\\"I‘m Ikun\\" class=\\"k-page-header__icon\\"><span class=\\"k-icon--base k-icon--base__dark k-icon-transition i-carbon-arrow-left\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 16px; height: 16px;\\"></span></div> <div class=\\"k-page-header__title\\">I‘m Ikun</div></div> <div class=\\"k-divider k-divider--vertical k-divider--vertical__solid k-divider--content-center b-ikun-bd-base\\" data-divider=\\"k-divider\\"></div> <div class=\\"k-page-header__content\\"></div></div> </div> </div>"`;

exports[`Test: KPageHeader > slots: breadcrumb 1`] = `"<div class=\\"k-page-header k-page-header--has-breadcrumb\\"><div class=\\"k-page-header__breadcrumb\\"><span slot=\\"breadcrumb\\">custom breadcrumb</span></div> <div class=\\"k-page-header__head\\"><div class=\\"k-page-header__left\\"><div class=\\"k-page-header__back\\" role=\\"button\\" tabindex=\\"0\\" aria-hidden=\\"true\\"><div aria-label=\\"Back\\" class=\\"k-page-header__icon\\"><span class=\\"k-icon--base k-icon--base__dark k-icon-transition i-carbon-arrow-left\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 16px; height: 16px;\\"></span></div> <div class=\\"k-page-header__title\\">Back</div></div> <div class=\\"k-divider k-divider--vertical k-divider--vertical__solid k-divider--content-center b-ikun-bd-base\\" data-divider=\\"k-divider\\"></div> <div class=\\"k-page-header__content\\"></div></div> </div> </div>"`;

exports[`Test: KPageHeader > slots: content 1`] = `"<div class=\\"k-page-header\\"> <div class=\\"k-page-header__head\\"><div class=\\"k-page-header__left\\"><div class=\\"k-page-header__back\\" role=\\"button\\" tabindex=\\"0\\" aria-hidden=\\"true\\"><div aria-label=\\"Back\\" class=\\"k-page-header__icon\\"><span class=\\"k-icon--base k-icon--base__dark k-icon-transition i-carbon-arrow-left\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 16px; height: 16px;\\"></span></div> <div class=\\"k-page-header__title\\">Back</div></div> <div class=\\"k-divider k-divider--vertical k-divider--vertical__solid k-divider--content-center b-ikun-bd-base\\" data-divider=\\"k-divider\\"></div> <div class=\\"k-page-header__content\\"><span slot=\\"content\\">custom content</span></div></div> </div> </div>"`;

exports[`Test: KPageHeader > slots: default 1`] = `"<div class=\\"k-page-header k-page-header--contentful\\"> <div class=\\"k-page-header__head\\"><div class=\\"k-page-header__left\\"><div class=\\"k-page-header__back\\" role=\\"button\\" tabindex=\\"0\\" aria-hidden=\\"true\\"><div aria-label=\\"Back\\" class=\\"k-page-header__icon\\"><span class=\\"k-icon--base k-icon--base__dark k-icon-transition i-carbon-arrow-left\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 16px; height: 16px;\\"></span></div> <div class=\\"k-page-header__title\\">Back</div></div> <div class=\\"k-divider k-divider--vertical k-divider--vertical__solid k-divider--content-center b-ikun-bd-base\\" data-divider=\\"k-divider\\"></div> <div class=\\"k-page-header__content\\"></div></div> </div> <div class=\\"k-page-header--main\\" data-pager-header=\\"main\\"><span>custom default</span></div></div>"`;

exports[`Test: KPageHeader > slots: extra 1`] = `"<div class=\\"k-page-header k-page-header--has-extra\\"> <div class=\\"k-page-header__head\\"><div class=\\"k-page-header__left\\"><div class=\\"k-page-header__back\\" role=\\"button\\" tabindex=\\"0\\" aria-hidden=\\"true\\"><div aria-label=\\"Back\\" class=\\"k-page-header__icon\\"><span class=\\"k-icon--base k-icon--base__dark k-icon-transition i-carbon-arrow-left\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 16px; height: 16px;\\"></span></div> <div class=\\"k-page-header__title\\">Back</div></div> <div class=\\"k-divider k-divider--vertical k-divider--vertical__solid k-divider--content-center b-ikun-bd-base\\" data-divider=\\"k-divider\\"></div> <div class=\\"k-page-header__content\\"></div></div> <div class=\\"k-page-header--extra\\"><span slot=\\"extra\\">custom extra</span></div></div> </div>"`;

exports[`Test: KPageHeader > slots: icon 1`] = `"<div class=\\"k-page-header\\"> <div class=\\"k-page-header__head\\"><div class=\\"k-page-header__left\\"><div class=\\"k-page-header__back\\" role=\\"button\\" tabindex=\\"0\\" aria-hidden=\\"true\\"><div aria-label=\\"Back\\" class=\\"k-page-header__icon\\"><span slot=\\"icon\\">custom icon</span></div> <div class=\\"k-page-header__title\\">Back</div></div> <div class=\\"k-divider k-divider--vertical k-divider--vertical__solid k-divider--content-center b-ikun-bd-base\\" data-divider=\\"k-divider\\"></div> <div class=\\"k-page-header__content\\"></div></div> </div> </div>"`;

exports[`Test: KPageHeader > slots: title 1`] = `"<div class=\\"k-page-header\\"> <div class=\\"k-page-header__head\\"><div class=\\"k-page-header__left\\"><div class=\\"k-page-header__back\\" role=\\"button\\" tabindex=\\"0\\" aria-hidden=\\"true\\"><div aria-label=\\"Back\\" class=\\"k-page-header__icon\\"><span class=\\"k-icon--base k-icon--base__dark k-icon-transition i-carbon-arrow-left\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 16px; height: 16px;\\"></span></div> <div class=\\"k-page-header__title\\"><span slot=\\"title\\">custom title</span></div></div> <div class=\\"k-divider k-divider--vertical k-divider--vertical__solid k-divider--content-center b-ikun-bd-base\\" data-divider=\\"k-divider\\"></div> <div class=\\"k-page-header__content\\"></div></div> </div> </div>"`;
